security-csrf/src/main/java/io/micronaut/security/csrf/filter/CsrfFilter.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright 2017-2024 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.security.csrf.filter;
+
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.core.order.Ordered;
+import io.micronaut.core.util.StringUtils;
+import io.micronaut.http.*;
+import io.micronaut.http.annotation.Filter;
+import io.micronaut.http.annotation.RequestFilter;
+import io.micronaut.http.filter.FilterPatternStyle;
+import io.micronaut.http.filter.HttpServerFilter;
+import io.micronaut.http.filter.ServerFilterChain;
+import io.micronaut.http.filter.ServerFilterPhase;
+import io.micronaut.http.server.exceptions.ExceptionHandler;
+import io.micronaut.security.authentication.Authentication;
+import io.micronaut.security.authentication.AuthorizationException;
+import io.micronaut.security.csrf.resolver.CsrfTokenResolver;
+import io.micronaut.security.csrf.resolver.ReactiveCsrfTokenResolver;
+import io.micronaut.security.csrf.validator.CsrfTokenValidator;
+import io.micronaut.security.filters.SecurityFilter;
+import org.reactivestreams.Publisher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+/**
+ * {@link RequestFilter} which validates CSRF tokens and rejects a request if the token is invalid.
+ * Which requests are intercepted can be controlled via {@link io.micronaut.security.csrf.CsrfConfiguration}.
+ * @author Sergio del Amo
+ * @since 4.11.0
+ */
+@Internal
+@Requires(property = CsrfFilterConfigurationProperties.PREFIX + ".enabled", value = StringUtils.TRUE, defaultValue = StringUtils.TRUE)
+@Requires(classes = { ExceptionHandler.class, HttpRequest.class })
+@Requires(beans = { CsrfTokenValidator.class })
+@Filter(patternStyle = FilterPatternStyle.REGEX,
+        value = "${" + CsrfFilterConfigurationProperties.PREFIX + ".regex-pattern:" + CsrfFilterConfigurationProperties.DEFAULT_REGEX_PATTERN + "}")
+final class CsrfFilter implements Ordered, HttpServerFilter {
```

Comment:
I think you can make this a method-based filter that returns a Publisher<Optional<HttpResponse<?>>>, with an empty optional if the request should pass. It should be much faster.

Reply:
I originally tried with a method-based filter but I was not able to work with `Publisher<HttpResponse>`and I [changed it to a HttpServerFilter](https://github.com/micronaut-projects/micronaut-security/pull/1826/commits/63a6c6be312947699037d1f363ade76d0770b938).

Reply:
I changed it https://github.com/micronaut-projects/micronaut-security/pull/1826/commits/00df7d64cc7c9a8583c6d091d3cb785634922234